// ts-nats/lib/databuffer.js
"use strict";
/*
 * Copyright 2018 The NATS Authors
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
var EMPTY_BUF = Buffer.allocUnsafe(0);
var CR = 13;
var LF = 10;
var ParserState;
(function (ParserState) {
    ParserState[ParserState["START"] = 0] = "START";
    ParserState[ParserState["CR"] = 1] = "CR";
})(ParserState || (ParserState = {}));
/**
 * @hidden
 */
var DataBuffer = /** @class */ (function () {
    function DataBuffer() {
        this.buffers = [];
        this.byteLength = 0;
    }
    DataBuffer.prototype.pack = function () {
        if (this.buffers.length > 1) {
            var v = Buffer.concat(this.buffers, this.byteLength);
            this.buffers = [v];
        }
    };
    DataBuffer.prototype.drain = function (n) {
        if (n === undefined) {
            n = this.byteLength;
        }
        if (this.byteLength >= n) {
            this.pack();
            var v = this.buffers.pop();
            if (v) {
                var d = v.slice(0, n);
                if (this.byteLength > n) {
                    this.buffers.push(v.slice(n));
                }
                this.byteLength -= n;
                return d;
            }
        }
        return EMPTY_BUF;
    };
    DataBuffer.prototype.fill = function (data) {
        if (data) {
            this.buffers.push(data);
            this.byteLength += data.byteLength;
        }
    };
    DataBuffer.prototype.protoLen = function () {
        var ps = ParserState.START;
        var offset = 0;
        for (var j = 0; j < this.buffers.length; j++) {
            var cb = this.buffers[j];
            for (var i = 0; i < cb.byteLength; i++) {
                var v = cb.readUInt8(i);
                switch (ps) {
                    case ParserState.START:
                        switch (v) {
                            case CR:
                                ps = ParserState.CR;
                                break;
                            default:
                        }
                        break;
                    case ParserState.CR:
                        switch (v) {
                            case LF:
                                // we want a length not an index
                                return offset + i + 1;
                            default:
                                ps = ParserState.START;
                        }
                        break;
                }
            }
            offset += cb.byteLength;
        }
        return -1;
    };
    DataBuffer.prototype.peek = function () {
        if (this.buffers.length) {
            this.pack();
            return this.buffers[0];
        }
        return EMPTY_BUF;
    };
    DataBuffer.prototype.size = function () {
        return this.byteLength;
    };
    DataBuffer.prototype.length = function () {
        return this.buffers.length;
    };
    DataBuffer.prototype.reset = function () {
        var a = this.buffers;
        this.buffers = [];
        this.byteLength = 0;
        return a;
    };
    return DataBuffer;
}());
exports.DataBuffer = DataBuffer;
