// ts-nats/lib/muxsubscriptions.js
"use strict";
/*
 * Copyright 2018 The NATS Authors
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
var util_1 = require("./util");
/**
 * @hidden
 */
var MuxSubscriptions = /** @class */ (function () {
    function MuxSubscriptions() {
        this.reqs = {};
        this.length = 0;
    }
    MuxSubscriptions.prototype.init = function () {
        this.baseInbox = util_1.createInbox() + ".";
        return this.baseInbox;
    };
    MuxSubscriptions.prototype.add = function (r) {
        if (!isNaN(r.received)) {
            r.received = 0;
        }
        this.length++;
        this.reqs[r.token] = r;
    };
    MuxSubscriptions.prototype.get = function (token) {
        if (token in this.reqs) {
            return this.reqs[token];
        }
        return null;
    };
    MuxSubscriptions.prototype.cancel = function (r) {
        if (r && r.timeout) {
            clearTimeout(r.timeout);
            delete r.timeout;
        }
        if (r.token in this.reqs) {
            delete this.reqs[r.token];
            this.length--;
        }
    };
    MuxSubscriptions.prototype.getToken = function (m) {
        var s = "";
        if (m) {
            s = m.subject || "";
        }
        if (s.indexOf(this.baseInbox) === 0) {
            return s.substring(this.baseInbox.length);
        }
        return null;
    };
    MuxSubscriptions.prototype.dispatcher = function () {
        var mux = this;
        return function (error, m) {
            var token = mux.getToken(m);
            if (token) {
                var r = mux.get(token);
                if (r) {
                    r.received++;
                    r.callback(error, m);
                    if (r.max && r.received >= r.max) {
                        mux.cancel(r);
                    }
                }
            }
        };
    };
    ;
    return MuxSubscriptions;
}());
exports.MuxSubscriptions = MuxSubscriptions;
